Examples/Algorithms/EventRecording/src/PrimaryGeneratorAction.cpp

```diff
@@ -0,0 +1,66 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "PrimaryGeneratorAction.hpp"
+#include <stdexcept>
+#include "G4Event.hh"
+#include "G4ParticleDefinition.hh"
+#include "G4ParticleGun.hh"
+#include "G4RandomDirection.hh"
+#include "G4UnitsTable.hh"
+#include "Randomize.hh"
+
+ActsExamples::PrimaryGeneratorAction*
+    ActsExamples::PrimaryGeneratorAction::s_instance = nullptr;
+
+ActsExamples::PrimaryGeneratorAction::PrimaryGeneratorAction(G4int randomSeed1,
+                                                             G4int randomSeed2)
+    : G4VUserPrimaryGeneratorAction(), m_particleGun(nullptr) {
+  // Configure the run
+  if (s_instance) {
+    throw std::logic_error("Attempted to duplicate a singleton");
+  } else {
+    s_instance = this;
+  }
+
+  // Configure the gun
+  G4int nofParticles = 1;
+  m_particleGun = std::make_unique<G4ParticleGun>(nofParticles);
+
+  // Prepare the particle table
+  m_particleTable = G4ParticleTable::GetParticleTable();
+
+  // Set the random seeds
+  CLHEP::HepRandom::getTheEngine()->setSeed(randomSeed1, randomSeed2);
+}
+
+ActsExamples::PrimaryGeneratorAction::~PrimaryGeneratorAction() {
+  s_instance = nullptr;
+}
+
+ActsExamples::PrimaryGeneratorAction*
+ActsExamples::PrimaryGeneratorAction::instance() {
+  // Static acces function via G4RunManager
+  return s_instance;
+}
+
+void ActsExamples::PrimaryGeneratorAction::prepareParticleGun(
+    G4int pdg, G4double momentum, G4ThreeVector pos, G4ThreeVector dir) {
```

Comment:
Consider taking a `const ActsFatras::Particle&` a.k.a. `const ActsExamples::SimParticle& as input and to perform the necessary conversion only within the function.

Reply:
done